### ISSUE TITLE
large scale gragh regression

### ISSUE THREAD

**Issue by ryuta0311** [pyg-team/pytorch_geometric#1661]
I have very large scale graphs.
Each graph represents a 3d object.
The number of graphs is 69 and each graph
includes 4000000 nodes and 12000000 edges.
I want to predict a scalar for each graph.

Now, I use a SAGPooling: Conv and Pool five times follwed by two layer mlp.
However, top k pooling takes a lot of time.

What kind of architecture fit this purpuse?

**Comment by tchaton** [comment]
Are you going mesh classification ? If yes, I would suggest to see them as pointcloud and use TorchPoints3d.

**Comment by ryuta0311** [comment]
No in this settings, I want to make a gragh regression.

**Comment by asadabbas09** [comment]
I'm also working on a similar problem but in addition to predicting scaler for each graph, I'm also trying to predict multiple scaler values for each node as well. I'm new to graphs, so not sure how to do edge and graph regression simultaneously. Any help would be appreciated.

**Comment by rusty1s** [comment]
You can return both node and graph predictions in your model's `forward` function, *e.g.*:
```python
x = self.gnn(x, edge_index)
out_node = self.node_mlp(x)

x = global_mean_pool(x, batch)
out_graph = self.graph_mlp(x)

return out_node, out_graph
```
and apply a regression loss both on `out_node` and `out_graph`, *e.g.*:
```python
loss = F.mse_loss(out_node, y_node) + F.mse_loss(out_graph, y_graph)
```

**Comment by asadabbas09** [comment]
@rusty1s  thanks for the quick response.

a quick follow up question. I'm struggling to understand how can I construct GCN to work with this dataset structure:

`Data(face=[3, 10664], pos=[5334, 3], y=[5334, 7])`

My understanding is that I will need to convert face to edges using: `transforms.FaceToEdge`.

Do I also need to convert `pos=[5334, 3]` using any other transform? 

Secondly, in addition to the target prediction for each node `y=[5334, 7]`, I have an extra scaler for that I would like to predict for the whole Graph `y = [12]`.

When building up the GCN, how should I define network structure, because I keep getting `size mismatch` error?

**Comment by rusty1s** [comment]
I think there are two ways to approach:
1. Use the edges of your mesh to perform convolution (via `FaceToEdge`)
2. Sample points on the mesh faces (via `SamplePoints`)

In case yo are using `FaceToEdge`, there is no need to convert `pos`, all your "nodes" are still there.

You may want to show me a small example that throws the `size mismatch` error, otherwise it's hard for me to give any advice :)

**Comment by asadabbas09** [comment]
I have following dataset structure:
```
Dataset: MyDataset(40):
====================
Number of graphs: 40

Data(face=[3, 10664], pos=[5334, 3], y=[5334, 7])
=============================================================
Number of nodes: 5334
Number of faces: 10664
```

I have followed the Point cloud classification collab notebook `SamplePoints` and use `knn_graph`.  I'm using `MSE` loss.

```
class PointNet(torch.nn.Module):
    def __init__(self):
        super(PointNet, self).__init__()

        torch.manual_seed(12345)
        self.conv1 = PointNetLayer(3, 32)
        self.conv2 = PointNetLayer(32, 32)
        self.classifier = Linear(32, dataset.num_classes)
        
    def forward(self, pos, batch):
        # Compute the kNN graph:
        # Here, we need to pass the batch vector to the function call in order
        # to prevent creating edges between points of different examples.
        # We also add `loop=True` which will add self-loops to the graph in
        # order to preserve central point information.
        edge_index = knn_graph(pos, k=16, batch=batch, loop=True)
        
        # 3. Start bipartite message passing.
        h = self.conv1(h=pos, pos=pos, edge_index=edge_index)
        h = h.relu()
        h = self.conv2(h=h, pos=pos, edge_index=edge_index)
        h = h.relu()

        # 4. Global Pooling.
        h = global_mean_pool(h, batch)  # [num_examples, hidden_channels]
        
        # 5. Classifier.
        return self.classifier(h)
```
I'm not sure how to construct the Graph model for node regression, as during loss computation I get following error:

```
print(np.shape(logits), np.shape(data.y)) # My batch size is 8
loss = criterion(logits, data.y)  # Loss computation.
```
Output:
```
torch.Size([8, 7]) torch.Size([42672, 7]) 
/usr/local/lib/python3.6/dist-packages/torch/nn/modules/loss.py:445: UserWarning:

Using a target size (torch.Size([42672, 7])) that is different to the input size (torch.Size([8, 7])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.
```

Can you help me with this one, and also let me know if there is a resource about how to use the edges of your mesh to perform convolution (via FaceToEdge).

Many Thanks!

**Comment by asadabbas09** [comment]
I think I have figured that out, it seems that during Sampling Points transformation, data structure was changed to

`pos=[128, 3], y=[5334, 7]`

So instead of using `SamplePoints` and `knn_graph`, instead  I used `Face2Edge` to get `edge_index`.  Training seems to be running fine. 

```
def train(model, optimizer, loader):
    model.train()
    
    total_loss = 0
    for data in train_loader:
        optimizer.zero_grad()  # Clear gradients.
        logits = model(data, data.batch)  # Forward pass.
        #logits = model(data.pos.float())
        #print(np.shape(logits), np.shape(data.y))
        loss = criterion(logits, data.y.float())  # Loss computation.
        loss.backward()  # Backward pass.
        optimizer.step()  # Update model parameters.
        total_loss += loss.item() * data.num_graphs

    return total_loss / len(train_loader.dataset)
```
However, my train loss from train function is very high. Am I doing something wrong here? please guide.
```
...
Epoch: 38, Loss: 26493.3414
Epoch: 39, Loss: 26239.3484
Epoch: 40, Loss: 26582.2013
Epoch: 41, Loss: 26282.4272
Epoch: 42, Loss: 26239.8622
``

**Comment by rusty1s** [comment]
That's actually pretty hard to say and can depend on many things such as learning rate or other hyperparams. It might also be a good idea to normalize your targets for easier training.
